src/test_utils/snapshot.rs

```diff
@@ -0,0 +1,279 @@
+use std::collections::HashMap;
+
+use crate::{participants::Participant, protocol::MessageData};
+
+/// Constructor for specific (sender, messages) pairs transmitted during a protocol run
+#[derive(Debug, PartialEq, Clone)]
+struct ReceivedMessageSnapshot {
+    from: Participant,
+    message: MessageData,
+}
+
+impl ReceivedMessageSnapshot {
+    fn new(from: Participant, message: MessageData) -> Self {
+        Self { from, message }
+    }
+}
+
+/// Registers a particular participant's view of the received messages
```

Comment:
Nit: "Registers" sounds confusing here since this is a data structure, not a behavior.

```suggestion
/// A participant's view of their received messages
```